src/experimental/patterns/ec2-app.ts

```diff
@@ -0,0 +1,132 @@
+ * Once all instances in the new ASG are healthy in the target group, CloudFormation will remove the current ASG,
+ * leaving one ASG provisioned.
+ *
+ * This pattern also updates the User Data, running some commands AFTER yours.
+ * These changes are wrapped in start and end marking comments.
+ *
+ * This pattern should improve the reliability of scaling events triggered during a deployment.
+ * Unlike in Riff-Raff's "autoscaling" deployment, scaling alarms are never suspended.
+ * TODO test scaling alarm behaviour.
+ *
+ * To update the application's code, a CloudFormation update must be triggered.
+ * The best way to do this is to include the build number in the application artifact.
+ * TODO test User Data includes a build number.
+ *
+ * NOTE: This pattern:
+ *  - Is NOT compatible with the "autoscaling" Riff-Raff deployment type.
+ *  - Your application should include a build number in its filename.
+ *    This value will change across builds, and therefore create a CloudFormation template difference to be deployed.
+ *  - Requires the AWS CLI and `cfn-signal` binaries to be available on the instance, and on the `PATH`.
+ *    AMIgo adds these via the `aws-tools` role.
+ *
+ * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html
+ */
+export class GuEc2AppExperimental extends GuEc2App {
+  constructor(scope: GuStack, props: GuEc2AppExperimentalProps) {
+    super(scope, { ...props, updatePolicy: UpdatePolicy.replacingUpdate() });
+
+    const { applicationPort } = props;
+    const { minimumInstances } = props.scaling;
+    const { region, stackId } = scope;
+    const { autoScalingGroup, targetGroup } = this;
+    const { userData, role } = autoScalingGroup;
+    const cfnAutoScalingGroup = autoScalingGroup.node.defaultChild as CfnAutoScalingGroup;
+
+    new Policy(scope, "AsgReplacingUpdatePolicy", {
+      statements: [
+        // Allow usage of command `cfn-signal`.
+        new PolicyStatement({
+          actions: ["cloudformation:SignalResource"],
+          effect: Effect.ALLOW,
+          resources: [stackId],
+        }),
+
+        /*
+        Allow usage of command `aws elbv2 describe-target-health`.
+        AWS Elastic Load Balancing does not support resource based policies, so the resource has to be `*` (any) here.
+        See https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_aws-services-that-work-with-iam.html
+         */
+        new PolicyStatement({
+          actions: ["elasticloadbalancing:DescribeTargetHealth"],
+          effect: Effect.ALLOW,
+          resources: ["*"],
+        }),
+      ],
+    }).attachToRole(role);
+
+    /*
+    `ec2metadata` is available via `cloud-utils` installed on all Canonical Ubuntu AMIs.
+    See https://github.com/canonical/cloud-utils.
+
+    `aws` is available via AMIgo baked AMIs.
+    See https://github.com/guardian/amigo/tree/main/roles/aws-tools.
+     */
+    userData.addCommands(
```

Comment:
I'd love us to move away from teams having to write the boiler plate of fetching their application from S3. We don't necessarily have to think about this at the same time, but if we're already changing artifacts around to include the build number, is this an opportunity to remove this as well at the same time, with only one breaking change?

I'd say Ideally userdata should be used by teams for appplication specific work if required, and we'd have everything needed for deploying and running a "normal" application in place.

Again, no need to tackle everything at once so if this is non-trivial then for sure let's consider it later.